File: 0036-valid-sudoku/0036-valid-sudoku.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        rows = set()
        cols = set()
        boxes = set()

        for i in range(9):
            for j in range(9):
                if board[i][j] == '.':
                    continue
                
                val = board[i][j]

                if ((i, val) in rows or
                    (j, val) in cols or
                    (i // 3, j // 3, val) in boxes
                ) : return False

                rows.add((i, val))
                cols.add((j, val))
                boxes.add((i // 3, j // 3, val))

        return True

File: 0036-valid-sudoku/0036-valid-sudoku.java
class Solution {
    public boolean isValidSudoku(char[][] board) {
        Map<Integer, Set<Character>> rows = new HashMap<>();
        Map<Integer, Set<Character>> cols = new HashMap<>();
        Map<Integer, Set<Character>> squares = new HashMap<>();

        for (int i = 0; i < board.length; i ++) {
            rows.putIfAbsent(i, new HashSet<>());
            cols.putIfAbsent(i, new HashSet<>());
            squares.putIfAbsent(i, new HashSet<>());
        }

        for (int r = 0; r < board.length; r ++) {
            for (int c = 0; c < board[0].length; c++) {
                if (board[r][c] == '.') continue;

                char num = board[r][c];
                int boxIndex = (r / 3) * 3 + (c / 3);

                if (rows.get(r).contains(num) || cols.get(c).contains(num) || squares.get(boxIndex).contains(num)) {
                    return false;
                }

                rows.get(r).add(num);
                cols.get(c).add(num);
                squares.get(boxIndex).add(num);
            }
        }

        return true;
    }
}